drm/vc4: hdmi: Add CSC for BT601/709/2020 limited and full range output

6by9 opened the pull request:
The HVS always composes in the RGB domain, but there is a colourspace
conversion block on the output to allow for sending YCbCr over the
HDMI interface.
The colourspace on that link is configurable via the "Colorspace"
property on the connector, and that updates the infoframes. There
is also selection of limited or full range based on the mode selected
or an override.

Add code to update the CSC as well so that the metadata matches the
image data.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/vc4: hdmi: Add CSC for BT601/709/2020 limited and full range output

The HVS always composes in the RGB domain, but there is a colourspace
conversion block on the output to allow for sending YCbCr over the
HDMI interface.
The colourspace on that link is configurable via the "Colorspace"
property on the connector, and that updates the infoframes. There
is also selection of limited or full range based on the mode selected
or an override.

Add code to update the CSC as well so that the metadata matches the
image data.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

popcornmix commented on drivers/gpu/drm/vc4/vc4_hdmi.c:
```diff
@@ -970,42 +961,107 @@ static const u16 vc5_hdmi_csc_full_rgb_unity[3][4] = {
-	{ 0x06ce, 0x16e3, 0x024f, 0x0000 },
+static const u16 vc5_hdmi_csc_full_rgb_to_yuv_bt2020[2][3][4] = {
+	{
+		/* Full range - unity */
```
unity?

6by9 replied:
Too much copy/paste.

6by9 commented:
Fixed up.

Just waiting on jc-kynesim to produce a test app that sets the plane colourspace appropriately, and then will confirm values on the analyser.

popcornmix commented on drivers/gpu/drm/vc4/vc4_hdmi.c:
```diff
@@ -970,42 +960,106 @@ static const u16 vc5_hdmi_csc_full_rgb_unity[3][4] = {
 
 /*
- * Conversion between Full Range RGB and Full Range YUV444 using the
+ * Conversion between Full Range RGB and Full Range YUV using the
```
No longer "Full Range" YUV

6by9 replied:
Fixed

6by9 commented:
Updated as there were some DRM_MODE_COLORIMETRY_* values that weren't explicitly mapped to any matrices.

6by9 commented:
@jc-kynesim

popcornmix commented:
It needs rebasing following rpi-5.15.y rebase. You can always merge from 6by9:rpi-5.15.y-dpi.

6by9 commented:
This is why I asked yesterday if we were ready to merge. There's little point in rebasing on every change to rpi-5.15.y if we aren't looking at merging it. You can always cherry-pick the top commits.

Rebased now.

6by9 committed:
dtoverlays: Connect the backlight to the pitft35 display

DRM will automatically handle the backlight with the display if
told about the linking, so link the two together.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

popcornmix commented:
I think we should merge this, as it fixes a clear bug. If any further tweaking of matrix is required it can be a follow up. @6by9 can you rebase?

6by9 commented:
Rebased as requested (despite what the Github PR says).

Github seems to have a couple of issues at present as it took several attempts to push the branch due to an "internal server error", and it's not wanting to accept my comment now.
